File: packages/budget-enforcer/python/src/budget_enforcer/storage/interface.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation

from __future__ import annotations

from abc import ABC, abstractmethod

from budget_enforcer.types import PendingCommit, SpendingEnvelope, Transaction


class BudgetStorage(ABC):
    """
    Minimal persistence contract for the budget enforcer.

    Implementors may back this with Redis, SQLite, Postgres, or any key-value
    store. The default MemoryStorage is suitable for single-process use and
    testing only — state is lost when the process exits.
    """

    # ─── Envelopes ────────────────────────────────────────────────────────────

    @abstractmethod
    def get_envelope(self, envelope_id: str) -> SpendingEnvelope | None:
        ...

    @abstractmethod
    def get_envelope_by_category(self, category: str) -> SpendingEnvelope | None:
        ...

    @abstractmethod
    def save_envelope(self, envelope: SpendingEnvelope) -> None:
        ...

    @abstractmethod
    def list_envelopes(self) -> list[SpendingEnvelope]:
        ...

    # ─── Transactions ─────────────────────────────────────────────────────────

    @abstractmethod
    def save_transaction(self, transaction: Transaction) -> None:
        ...

    @abstractmethod
    def list_transactions(self) -> list[Transaction]:
        ...

    # ─── Pending commits ──────────────────────────────────────────────────────

    @abstractmethod
    def save_commit(self, commit: PendingCommit) -> None:
        ...

    @abstractmethod
    def get_commit(self, commit_id: str) -> PendingCommit | None:
        ...

    @abstractmethod
    def delete_commit(self, commit_id: str) -> None:
        ...

    @abstractmethod
    def list_commits(self) -> list[PendingCommit]:
        ...

File: packages/budget-enforcer/typescript/src/storage/interface.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

import type { SpendingEnvelope, Transaction, PendingCommit } from '../types.js';

/**
 * Minimal persistence contract for the budget enforcer.
 * Implementors may back this with Redis, SQLite, Postgres, or any KV store.
 */
export interface BudgetStorage {
  // Envelopes
  getEnvelope(id: string): Promise<SpendingEnvelope | null>;
  getEnvelopeByCategory(category: string): Promise<SpendingEnvelope | null>;
  saveEnvelope(envelope: SpendingEnvelope): Promise<void>;
  listEnvelopes(): Promise<readonly SpendingEnvelope[]>;

  // Transactions
  saveTransaction(transaction: Transaction): Promise<void>;
  listTransactions(): Promise<readonly Transaction[]>;

  // Pending commits (pre-authorizations)
  saveCommit(commit: PendingCommit): Promise<void>;
  getCommit(id: string): Promise<PendingCommit | null>;
  deleteCommit(id: string): Promise<void>;
  listCommits(): Promise<readonly PendingCommit[]>;
}
